---
title: "TSV_creation"
author: "Victor Billon"
date: "2024-06-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Only Part III and IV need to be ran with new datasets.

# Part I : Creating annotation
Producing 
##making windows for transcripts

###Protein coding transcripts
```{bash}
#Annotation
#/Users/victor/Documents/DATA/annotation/V43
#gencode.v43.basic.annotation.gtf

grep -w transcript *.gtf | grep -w MANE_Select > MANE_Select.protein_coding.gtf
awk 'OFS="\t" {print $0}' MANE_Select.protein_coding.gtf | tr -d '";' | sort -k1,1 -k2,2n  | awk -F \t -v OFS='\t' '{print $0}' | awk -v OFS="\t" '{ print $1,$4,$5,$12,$16,$7}' > MANE_Select.protein_coding.bed
##this still contains PAR genes
```

###lncRNA transcripts
```{bash}
grep -w transcript gencode.v43.basic.annotation.gtf | grep -w lncRNA | grep -w Ensembl_canonical | grep -v not_best_in_genome_evidence | grep -v 'transcript_support_level "5"' | grep -v 'transcript_support_level "4"' > Ensembl_canonical_TSL123.lncRNA.gtf
awk 'OFS="\t" {print $0}' Ensembl_canonical_TSL123.lncRNA.gtf | tr -d '";' | sort -k1,1 -k2,2n  | awk -F \t -v OFS='\t' '{print $0}' | awk -v OFS="\t" '{ print $1,$4,$5,$12,$16,$7}' > Ensembl_canonical_TSL123.lncRNA.bed
```

# Part II : Removing Blacklist from annotation
This part produces the windows per transcripts
### Protein Coding
```{zsh engine.opts='-i'}
ANNOTATION="/Users/victor/Documents/DATA/annotation/V43/MANE_Select.protein_coding.bed"
WORKING="/Users/victor/Documents/DATA/annotation/V43"
blacklist="/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed"
window=200

mkdir $WORKING/makewindow
awk -F "\t" -v OFS="\t" '{print $1,$2,$3,$4"_"$5"_"$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v  | bedtools makewindows -n $window -i srcwinnum -b stdin | sort -k1,1 -k2,2n > $WORKING/makewindow/v43.MANE_protein.window${window}.bed  

echo $WORKING/makewindow/v43.MANE_protein.window${window}.bed 
``` 

### lncRNA
```{zsh engine.opts='-i'}
ANNOTATION="/Users/victor/Documents/DATA/annotation/V43/Ensembl_canonical_TSL123.lncRNA.bed"
WORKING="/Users/victor/Documents/DATA/annotation/V43"
blacklist="/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed"
window=200


awk -F "\t" -v OFS="\t" '{print $1,$2,$3,$4"_"$5"_"$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v  | bedtools makewindows -n $window -i srcwinnum -b stdin | sort -k1,1 -k2,2n > $WORKING/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed 

echo $WORKING/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed 
```



# Part III : Associating score to bed
Input needed: the transcrippts with the window.

This is the part that people must do to get score per window. then Part IV to create the needed output
Part III is bash code, part IV is R code.


### Protein coding genes
First plan was to use bedtools map -o sum but the sum is not weighted so I'll use bedops wmean instead which is weighted. 
To work the bedgraph needs to get minus and plus in the name.
The zsh script below has to be written in a file WITHOUT the comments. and ran has follow :

chmod +x scoring.zsh
nohup zsh scoring.zsh
you can found and edit it here : 
~/INTRANET/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255
/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255
```{zsh engine.opts='-i'}
#!/usr/bin/zsh

## DO NOT FORGET TO CHANGE THE EXTENSION OF THE BEDGRAPH IF NEEDED
ext="bg"
##it works and validated by looking at head and tail of documents and exemple genes
WORKING="/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255"
window=200
######
umapk50="/Users/victor/Documents/DATA/Mappability/hg38/k50_score/highMap/k50.umap.hg38.0.8.bed" # hg38 mappability windows with mapp k50 > 80%
windowS="/Users/victor/Documents/DATA/annotation/V43/makewindow/v43.MANE_protein.window${window}.bed" #genes with the window
blacklist="/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed"
ANNOTATION="/Users/victor/Documents/DATA/annotation/V43/MANE_Select.protein_coding.bed"
#######

mkdir $WORKING/withzeros
mkdir $WORKING/mapHigh

for file in $WORKING/*.$ext ;
do
filename=$(basename "$file" .$ext);
echo "starting file :"
echo $filename;


if echo $filename | egrep -q "reverse|minus" ;  then  
  strand="-"
elif echo $filename | egrep -q "forward|plus" ; then
  strand='+'
fi

echo "removing blacklist region"
 bedtools intersect \
-a $WORKING/${filename}.$ext \
-b <( awk -F "\t" -v OFS="\t" -v myvar=$strand '{if ($6==myvar) print $1,$2,$3,$4"_"$5"_"$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v) \
 | sort -k1,1 -k2,2n > $WORKING/withzeros/${filename}.nonzeros.$ext


echo "removing low mappability region"


bedtools intersect -a $WORKING/withzeros/${filename}.nonzeros.${ext} -b $umapk50 -sorted | awk -F "\t" -v OFS="\t" '{print $1,$2,$3,".",$4}' > $WORKING/mapHigh/${filename}.0.8.$ext

echo "scoring windows"

bedmap --echo --wmean --delim "\t" $windowS $WORKING/mapHigh/${filename}.0.8.$ext | awk -F "_" -v OFS="\t" '{print $1,$2,$3,$4}' | awk -F "\t" -v OFS="\t" -v name="$filename" '{ print $0,$4"_"$5"_"$6"_"$7,name}' | awk -F "\t" -v OFS="\t" '{ print "protein-coding",$1,$2,$3,$4,$5,$6,$7,$9,$10,$8 }' > $WORKING/${filename}.window${window}.MANE.wmean.name.score ;

echo "done"

done

rm -r $WORKING/withzeros
rm -r $WORKING/mapHigh
mkdir $WORKING/protein_coding_score
mv $WORKING/*.score $WORKING/protein_coding_score/

``` 

### For long non coding RNA :
chmod +x lncRNA.zsh
nohup zsh lncRNA.zsh &> lncRNA.nohup
```{zsh engine.opts='-i'}
#!/usr/bin/zsh

ext="bg"
WORKING="/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255"
window=200

umapk50="/Users/victor/Documents/DATA/Mappability/hg38/k50_score/highMap/k50.umap.hg38.0.8.bed"
windowS="/Users/victor/Documents/DATA/annotation/V43/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed"
blacklist="/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed"
ANNOTATION="/Users/victor/Documents/DATA/annotation/V43/Ensembl_canonical_TSL123.lncRNA.bed"

mkdir $WORKING/withzeros
mkdir $WORKING/mapHigh

for file in $WORKING/*.$ext ;
do
filename=$(basename "$file" .$ext) ;
echo "starting file :"
echo $filename ;

if echo $filename | egrep -q "reverse|minus" ;  then 
  strand="-"
elif echo $filename | egrep -q "forward|plus" ; then
  strand='+'
fi


echo "removing blacklist region"
 bedtools intersect \
-a $WORKING/${filename}.$ext \
-b <( awk -F "\t" -v OFS="\t" -v myvar=$strand '{if ($6==myvar) print $1,$2,$3,$4"_"$5"_"$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v) \
 | sort -k1,1 -k2,2n > $WORKING/withzeros/${filename}.nonzeros.$ext


echo "removing low mappability region"

bedtools intersect -a $WORKING/withzeros/${filename}.nonzeros.$ext -b $umapk50 -sorted | awk -F "\t" -v OFS="\t" '{print $1,$2,$3,".",$4}' > $WORKING/mapHigh/${filename}.0.8.$ext

echo "scoring windows"

bedmap --echo --wmean --delim "\t" $windowS $WORKING/mapHigh/${filename}.0.8.$ext | awk -F "_" -v OFS="\t" '{print $1,$2,$3,$4}' | awk -F "\t" -v OFS="\t" -v name="$filename" '{ print $0,$4"_"$5"_"$6"_"$7,name}' | awk -F "\t" -v OFS="\t" '{ print "lncRNA",$1,$2,$3,$4,$5,$6,$7,$9,$10,$8 }' > $WORKING/${filename}.window${window}.MANE.wmean.name.score ;

echo "done"

done

rm -r $WORKING/withzeros
rm -r $WORKING/mapHigh

mkdir $WORKING/lncRNA_score
mv $WORKING/*.score $WORKING/lncRNA_score/

```






# Part IV : Joining tables
### Protein coding genes
```{r}
main_directory <- "/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255"
write_file_protein_coding <- paste0(main_directory,"/protCoding_dTAG_Cugusi_stranded_20230810.tsv")
write_file_lncRNA <- paste0(main_directory,"/lncRNA_dTAG_Cugusi_stranded_20230810.tsv")
Big_tsv <-paste0(main_directory,"/dTAG_Cugusi_stranded_20230810.tsv")  
```

```{r}
main_directory <- "/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Xiong2021/analysis_KD/bedgraph"
write_file_protein_coding <- paste0(main_directory,"/protCoding_XiongsiSAFB_stranded_20240214.tsv")
write_file_lncRNA <- paste0(main_directory,"/lncRNA_XiongsiSAFB_stranded_20240214.tsv")
Big_tsv <-paste0(main_directory,"/XiongsiSAFB_stranded_20240214.tsv")  
```

```{r}
main_directory <- "/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/SousaLuis2021/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/stranded/bedgraph255"
write_file_protein_coding <- paste0(main_directory,"/protCoding_SousaLuis_stranded_20240322.tsv")
write_file_lncRNA <- paste0(main_directory,"/lncRNA_SousaLuis_stranded_20240322.tsv")
Big_tsv <-paste0(main_directory,"/SousaLuis_stranded_20240322.tsv")  
```

~5 min
```{r}
library(dplyr)
library(purrr)

### DO NOT FORGET TO CHANGE THE BEDGRAPH EXTENSION TO TH ONE IN USE IN THE FOLDER
window <- 200

working_directory <- paste0(main_directory,"/protein_coding_score")
bedgraph_files <- list.files(main_directory, pattern = "*.bg", full.names = TRUE)

files <- bedgraph_files %>%
  map(~{
    filename <- tools::file_path_sans_ext(basename(.))
    file.path(working_directory, paste0(filename, ".window", window, ".MANE.wmean.name.score"))
  })
# reading all the files
list_of_dfs <- lapply(files, read.delim, header= F, sep= "\t",  na.strings = "NAN", dec = ".", col.names = c("biotype","chr", "coor1", "coor2","transcript", "gene", "strand","window","id","dataset","score"), stringsAsFactors=FALSE)
# joining all the files
joined_df <- purrr::reduce(list_of_dfs, dplyr::left_join, by = c("biotype","chr","coor1","coor2","transcript","gene","strand","window","id")) %>% 
   dplyr::filter(strand!="Y") ## the last filter remove the PAR genes (pseudoautosomal genes both in X and Y)
 
 
 write.table(joined_df, file = write_file_protein_coding, sep = "\t", row.names = FALSE, col.names = FALSE, quote = F)
 rm(list_of_dfs)
 rm(joined_df)
```

### lncRNA

```{r}
library(dplyr)
library(purrr)

### DO NOT FORGET TO CHANGE THE BEDGRAPH EXTENSION TO TH ONE IN USE IN THE FOLDER
window <- 200
working_directory <- paste0(main_directory,"/lncRNA_score")

bedgraph_files <- list.files(main_directory, pattern = "*.bg", full.names = TRUE)

files <- bedgraph_files %>%
  map(~{
    filename <- tools::file_path_sans_ext(basename(.))
    file.path(working_directory, paste0(filename, ".window", window, ".MANE.wmean.name.score"))
  })
# reading all the files
list_of_dfs <- lapply(files, read.delim, header= F, sep= "\t",  na.strings = "NAN", dec = ".", col.names = c("biotype","chr", "coor1", "coor2","transcript", "gene", "strand","window","id","dataset","score"), stringsAsFactors=FALSE)
# joining all the files
joined_df <- purrr::reduce(list_of_dfs, dplyr::left_join, by = c("biotype","chr","coor1","coor2","transcript","gene","strand","window","id")) %>% 
   dplyr::filter(strand!="Y") ## the last filter remove the PAR genes (pseudoautosomal genes both in X and Y)
 
 
 write.table(joined_df, file = write_file_lncRNA, sep = "\t", row.names = FALSE, col.names = FALSE, quote = F)

  rm(list_of_dfs)
 rm(joined_df)
```



## Merging the two files : 

```{r}
library(dplyr)

tsv_files <- list.files(main_directory, pattern = "*.tsv", full.names = TRUE)

files <- tsv_files %>%
  map(~{
    filename <- tools::file_path_sans_ext(basename(.))
    file.path(main_directory, paste0(filename, ".tsv"))
  })
# reading all the files
list_of_dfs <- lapply(files, read.delim, header= F, sep= "\t",  na.strings = "NAN", dec = ".", stringsAsFactors=FALSE)
# joining all the files
bound_df <- purrr::reduce(list_of_dfs, dplyr::bind_rows)
 

write.table(bound_df, file = Big_tsv, sep = "\t", row.names = FALSE, col.names = FALSE, quote = F)
rm(list_of_dfs)
rm(bound_df)
```

