<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>TSV_creation</title>
<style type="text/css">
/**
 * Prism.s theme ported from highlight.js's xcode style
 */
pre code {
  padding: 1em;
}
.token.comment {
  color: #007400;
}
.token.punctuation {
  color: #999;
}
.token.tag,
.token.selector {
  color: #aa0d91;
}
.token.boolean,
.token.number,
.token.constant,
.token.symbol {
  color: #1c00cf;
}
.token.property,
.token.attr-name,
.token.string,
.token.char,
.token.builtin {
  color: #c41a16;
}
.token.inserted {
  background-color: #ccffd8;
}
.token.deleted {
  background-color: #ffebe9;
}
.token.operator,
.token.entity,
.token.url,
.language-css .token.string,
.style .token.string {
  color: #9a6e3a;
}
.token.atrule,
.token.attr-value,
.token.keyword {
  color: #836c28;
}
.token.function,
.token.class-name {
  color: #DD4A68;
}
.token.regex,
.token.important,
.token.variable {
  color: #5c2699;
}
.token.important,
.token.bold {
  font-weight: bold;
}
.token.italic {
  font-style: italic;
}
</style>
<style type="text/css">
body {
  font-family: sans-serif;
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 1.5;
  box-sizing: border-box;
}
body, .footnotes, code { font-size: .9em; }
li li { font-size: .95em; }
*, *:before, *:after {
  box-sizing: inherit;
}
pre, img { max-width: 100%; }
pre, pre:hover {
  white-space: pre-wrap;
  word-break: break-all;
}
pre code {
  display: block;
  overflow-x: auto;
}
code { font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace; }
:not(pre) > code, code[class] { background-color: #F8F8F8; }
code.language-undefined, pre > code:not([class]) {
  background-color: inherit;
  border: 1px solid #eee;
}
table {
  margin: auto;
  border-top: 1px solid #666;
}
table thead th { border-bottom: 1px solid #ddd; }
th, td { padding: 5px; }
thead, tfoot, tr:nth-child(even) { background: #eee; }
blockquote {
  color: #666;
  margin: 0;
  padding-left: 1em;
  border-left: 0.5em solid #eee;
}
hr, .footnotes::before { border: 1px dashed #ddd; }
.frontmatter { text-align: center; }
#TOC .numbered li { list-style: none; }
#TOC .numbered { padding-left: 0; }
#TOC .numbered ul { padding-left: 1em; }
table, .body h2 { border-bottom: 1px solid #666; }
.body .appendix, .appendix ~ h2 { border-bottom-style: dashed; }
.footnote-ref a::before { content: "["; }
.footnote-ref a::after { content: "]"; }
section.footnotes::before {
  content: "";
  display: block;
  max-width: 20em;
}

@media print {
  body {
    font-size: 12pt;
    max-width: 100%;
  }
  tr, img { page-break-inside: avoid; }
}
@media only screen and (min-width: 992px) {
  pre { white-space: pre; }
}
</style>
</head>
<body>
<div class="frontmatter">
<div class="title"><h1>TSV_creation</h1></div>
<div class="author"><h2>Victor Billon</h2></div>
<div class="date"><h3>2024-06-17</h3></div>
</div>
<div class="body">
<p>Only Part III and IV need to be ran with new datasets.</p>
<h1 id="part-i-creating-annotation">Part I : Creating annotation</h1>
<p>Producing
##making windows for transcripts</p>
<p>###Protein coding transcripts</p>
<pre><code class="language-bash">#Annotation
#/Users/victor/Documents/DATA/annotation/V43
#gencode.v43.basic.annotation.gtf

grep -w transcript *.gtf | grep -w MANE_Select &gt; MANE_Select.protein_coding.gtf
awk 'OFS=&quot;\t&quot; {print $0}' MANE_Select.protein_coding.gtf | tr -d '&quot;;' | sort -k1,1 -k2,2n  | awk -F \t -v OFS='\t' '{print $0}' | awk -v OFS=&quot;\t&quot; '{ print $1,$4,$5,$12,$16,$7}' &gt; MANE_Select.protein_coding.bed
##this still contains PAR genes
</code></pre>
<pre><code>## grep: *.gtf: No such file or directory
</code></pre>
<p>###lncRNA transcripts</p>
<pre><code class="language-bash">grep -w transcript gencode.v43.basic.annotation.gtf | grep -w lncRNA | grep -w Ensembl_canonical | grep -v not_best_in_genome_evidence | grep -v 'transcript_support_level &quot;5&quot;' | grep -v 'transcript_support_level &quot;4&quot;' &gt; Ensembl_canonical_TSL123.lncRNA.gtf
awk 'OFS=&quot;\t&quot; {print $0}' Ensembl_canonical_TSL123.lncRNA.gtf | tr -d '&quot;;' | sort -k1,1 -k2,2n  | awk -F \t -v OFS='\t' '{print $0}' | awk -v OFS=&quot;\t&quot; '{ print $1,$4,$5,$12,$16,$7}' &gt; Ensembl_canonical_TSL123.lncRNA.bed
</code></pre>
<pre><code>## grep: gencode.v43.basic.annotation.gtf: No such file or directory
</code></pre>
<h1 id="part-ii-removing-blacklist-from-annotation">Part II : Removing Blacklist from annotation</h1>
<p>This part produces the windows per transcripts</p>
<h3 id="protein-coding">Protein Coding</h3>
<pre><code class="language-zsh">ANNOTATION=&quot;/Users/victor/Documents/DATA/annotation/V43/MANE_Select.protein_coding.bed&quot;
WORKING=&quot;/Users/victor/Documents/DATA/annotation/V43&quot;
blacklist=&quot;/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed&quot;
window=200

mkdir $WORKING/makewindow
awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; '{print $1,$2,$3,$4&quot;_&quot;$5&quot;_&quot;$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v  | bedtools makewindows -n $window -i srcwinnum -b stdin | sort -k1,1 -k2,2n &gt; $WORKING/makewindow/v43.MANE_protein.window${window}.bed  

echo $WORKING/makewindow/v43.MANE_protein.window${window}.bed 
</code></pre>
<pre><code>##                                
##  ___  ___ _ __   ___  ___ __ _ 
## / __|/ _ \ '_ \ / _ \/ __/ _` |
## \__ \  __/ | | |  __/ (_| (_| |
## |___/\___|_| |_|\___|\___\__,_|
##                                
## 
## Please think about your colleagues and submit long lasting jobs to the cluster.
## 
## /tmp - limited [1mlocal(B[m space, very temporary and cleaned up regularly
## /tmpdata - very fast [1mlocal(B[m space, no backups, no guarantees
## /scratch - fast [1mcluster(B[m space, no backups, no guarantees
## 
## Find a tool: module spider &lt;search_term&gt;
## Load a tool: module load &lt;module_name&gt;
## 
## Should you have any questions, please contact GBCS (gbcs@embl.de)
## or have a look here: https://gbcs.embl.de
## 
## mkdir: cannot create directory â€˜/Users/victor/Documents/DATA/annotation/V43/makewindowâ€™: No such file or directory
## zsh:7: command not found: bedtools
## awk: fatal: cannot open file `/Users/victor/Documents/DATA/annotation/V43/MANE_Select.protein_coding.bed' for reading (No such file or directory)
## zsh:7: command not found: bedtools
## zsh:7: no such file or directory: /Users/victor/Documents/DATA/annotation/V43/makewindow/v43.MANE_protein.window200.bed
## /Users/victor/Documents/DATA/annotation/V43/makewindow/v43.MANE_protein.window200.bed
</code></pre>
<h3 id="lncrna_1">lncRNA</h3>
<pre><code class="language-zsh">ANNOTATION=&quot;/Users/victor/Documents/DATA/annotation/V43/Ensembl_canonical_TSL123.lncRNA.bed&quot;
WORKING=&quot;/Users/victor/Documents/DATA/annotation/V43&quot;
blacklist=&quot;/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed&quot;
window=200


awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; '{print $1,$2,$3,$4&quot;_&quot;$5&quot;_&quot;$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v  | bedtools makewindows -n $window -i srcwinnum -b stdin | sort -k1,1 -k2,2n &gt; $WORKING/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed 

echo $WORKING/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed 
</code></pre>
<pre><code>##                                
##  ___  ___ _ __   ___  ___ __ _ 
## / __|/ _ \ '_ \ / _ \/ __/ _` |
## \__ \  __/ | | |  __/ (_| (_| |
## |___/\___|_| |_|\___|\___\__,_|
##                                
## 
## Please think about your colleagues and submit long lasting jobs to the cluster.
## 
## /tmp - limited [1mlocal(B[m space, very temporary and cleaned up regularly
## /tmpdata - very fast [1mlocal(B[m space, no backups, no guarantees
## /scratch - fast [1mcluster(B[m space, no backups, no guarantees
## 
## Find a tool: module spider &lt;search_term&gt;
## Load a tool: module load &lt;module_name&gt;
## 
## Should you have any questions, please contact GBCS (gbcs@embl.de)
## or have a look here: https://gbcs.embl.de
## 
## zsh:7: command not found: bedtools
## zsh:7: command not found: bedtools
## zsh:7: no such file or directory: /Users/victor/Documents/DATA/annotation/V43/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed
## awk: fatal: cannot open file `/Users/victor/Documents/DATA/annotation/V43/Ensembl_canonical_TSL123.lncRNA.bed' for reading (No such file or directory)
## /Users/victor/Documents/DATA/annotation/V43/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed
</code></pre>
<h1 id="part-iii-associating-score-to-bed">Part III : Associating score to bed</h1>
<p>Input needed: the transcrippts with the window.</p>
<p>This is the part that people must do to get score per window. then Part IV to create the needed output
Part III is bash code, part IV is R code.</p>
<h3 id="protein-coding-genes_1">Protein coding genes</h3>
<p>First plan was to use bedtools map -o sum but the sum is not weighted so Iâ€™ll use bedops wmean instead which is weighted.
To work the bedgraph needs to get minus and plus in the name.
The zsh script below has to be written in a file WITHOUT the comments. and ran has follow :</p>
<p>chmod +x scoring.zsh
nohup zsh scoring.zsh
you can found and edit it here :
~/INTRANET/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255
/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255</p>
<pre><code class="language-zsh">#!/usr/bin/zsh

## DO NOT FORGET TO CHANGE THE EXTENSION OF THE BEDGRAPH IF NEEDED
ext=&quot;bg&quot;
##it works and validated by looking at head and tail of documents and exemple genes
WORKING=&quot;/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255&quot;
window=200
######
umapk50=&quot;/Users/victor/Documents/DATA/Mappability/hg38/k50_score/highMap/k50.umap.hg38.0.8.bed&quot; # hg38 mappability windows with mapp k50 &gt; 80%
windowS=&quot;/Users/victor/Documents/DATA/annotation/V43/makewindow/v43.MANE_protein.window${window}.bed&quot; #genes with the window
blacklist=&quot;/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed&quot;
ANNOTATION=&quot;/Users/victor/Documents/DATA/annotation/V43/MANE_Select.protein_coding.bed&quot;
#######

mkdir $WORKING/withzeros
mkdir $WORKING/mapHigh

for file in $WORKING/*.$ext ;
do
filename=$(basename &quot;$file&quot; .$ext);
echo &quot;starting file :&quot;
echo $filename;


if echo $filename | egrep -q &quot;reverse|minus&quot; ;  then  
  strand=&quot;-&quot;
elif echo $filename | egrep -q &quot;forward|plus&quot; ; then
  strand='+'
fi

echo &quot;removing blacklist region&quot;
 bedtools intersect \
-a $WORKING/${filename}.$ext \
-b &lt;( awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; -v myvar=$strand '{if ($6==myvar) print $1,$2,$3,$4&quot;_&quot;$5&quot;_&quot;$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v) \
 | sort -k1,1 -k2,2n &gt; $WORKING/withzeros/${filename}.nonzeros.$ext


echo &quot;removing low mappability region&quot;


bedtools intersect -a $WORKING/withzeros/${filename}.nonzeros.${ext} -b $umapk50 -sorted | awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; '{print $1,$2,$3,&quot;.&quot;,$4}' &gt; $WORKING/mapHigh/${filename}.0.8.$ext

echo &quot;scoring windows&quot;

bedmap --echo --wmean --delim &quot;\t&quot; $windowS $WORKING/mapHigh/${filename}.0.8.$ext | awk -F &quot;_&quot; -v OFS=&quot;\t&quot; '{print $1,$2,$3,$4}' | awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; -v name=&quot;$filename&quot; '{ print $0,$4&quot;_&quot;$5&quot;_&quot;$6&quot;_&quot;$7,name}' | awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; '{ print &quot;protein-coding&quot;,$1,$2,$3,$4,$5,$6,$7,$9,$10,$8 }' &gt; $WORKING/${filename}.window${window}.MANE.wmean.name.score ;

echo &quot;done&quot;

done

rm -r $WORKING/withzeros
rm -r $WORKING/mapHigh
mkdir $WORKING/protein_coding_score
mv $WORKING/*.score $WORKING/protein_coding_score/

</code></pre>
<pre><code>##                                
##  ___  ___ _ __   ___  ___ __ _ 
## / __|/ _ \ '_ \ / _ \/ __/ _` |
## \__ \  __/ | | |  __/ (_| (_| |
## |___/\___|_| |_|\___|\___\__,_|
##                                
## 
## Please think about your colleagues and submit long lasting jobs to the cluster.
## 
## /tmp - limited [1mlocal(B[m space, very temporary and cleaned up regularly
## /tmpdata - very fast [1mlocal(B[m space, no backups, no guarantees
## /scratch - fast [1mcluster(B[m space, no backups, no guarantees
## 
## Find a tool: module spider &lt;search_term&gt;
## Load a tool: module load &lt;module_name&gt;
## 
## Should you have any questions, please contact GBCS (gbcs@embl.de)
## or have a look here: https://gbcs.embl.de
## 
## mkdir: cannot create directory â€˜/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255/withzerosâ€™: No such file or directory
## mkdir: cannot create directory â€˜/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255/mapHighâ€™: No such file or directory
## zsh:18: no matches found: /Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255/*.bg
</code></pre>
<h3 id="for-long-non-coding-rna">For long non coding RNA :</h3>
<p>chmod +x lncRNA.zsh
nohup zsh lncRNA.zsh &amp;&gt; lncRNA.nohup</p>
<pre><code class="language-zsh">#!/usr/bin/zsh

ext=&quot;bg&quot;
WORKING=&quot;/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255&quot;
window=200

umapk50=&quot;/Users/victor/Documents/DATA/Mappability/hg38/k50_score/highMap/k50.umap.hg38.0.8.bed&quot;
windowS=&quot;/Users/victor/Documents/DATA/annotation/V43/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed&quot;
blacklist=&quot;/Users/victor/Documents/DATA/Repeatmasker/blacklist/hg38-blacklist.v2.sorted.bed&quot;
ANNOTATION=&quot;/Users/victor/Documents/DATA/annotation/V43/Ensembl_canonical_TSL123.lncRNA.bed&quot;

mkdir $WORKING/withzeros
mkdir $WORKING/mapHigh

for file in $WORKING/*.$ext ;
do
filename=$(basename &quot;$file&quot; .$ext) ;
echo &quot;starting file :&quot;
echo $filename ;

if echo $filename | egrep -q &quot;reverse|minus&quot; ;  then 
  strand=&quot;-&quot;
elif echo $filename | egrep -q &quot;forward|plus&quot; ; then
  strand='+'
fi


echo &quot;removing blacklist region&quot;
 bedtools intersect \
-a $WORKING/${filename}.$ext \
-b &lt;( awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; -v myvar=$strand '{if ($6==myvar) print $1,$2,$3,$4&quot;_&quot;$5&quot;_&quot;$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v) \
 | sort -k1,1 -k2,2n &gt; $WORKING/withzeros/${filename}.nonzeros.$ext


echo &quot;removing low mappability region&quot;

bedtools intersect -a $WORKING/withzeros/${filename}.nonzeros.$ext -b $umapk50 -sorted | awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; '{print $1,$2,$3,&quot;.&quot;,$4}' &gt; $WORKING/mapHigh/${filename}.0.8.$ext

echo &quot;scoring windows&quot;

bedmap --echo --wmean --delim &quot;\t&quot; $windowS $WORKING/mapHigh/${filename}.0.8.$ext | awk -F &quot;_&quot; -v OFS=&quot;\t&quot; '{print $1,$2,$3,$4}' | awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; -v name=&quot;$filename&quot; '{ print $0,$4&quot;_&quot;$5&quot;_&quot;$6&quot;_&quot;$7,name}' | awk -F &quot;\t&quot; -v OFS=&quot;\t&quot; '{ print &quot;lncRNA&quot;,$1,$2,$3,$4,$5,$6,$7,$9,$10,$8 }' &gt; $WORKING/${filename}.window${window}.MANE.wmean.name.score ;

echo &quot;done&quot;

done

rm -r $WORKING/withzeros
rm -r $WORKING/mapHigh

mkdir $WORKING/lncRNA_score
mv $WORKING/*.score $WORKING/lncRNA_score/

</code></pre>
<pre><code>##                                
##  ___  ___ _ __   ___  ___ __ _ 
## / __|/ _ \ '_ \ / _ \/ __/ _` |
## \__ \  __/ | | |  __/ (_| (_| |
## |___/\___|_| |_|\___|\___\__,_|
##                                
## 
## Please think about your colleagues and submit long lasting jobs to the cluster.
## 
## /tmp - limited [1mlocal(B[m space, very temporary and cleaned up regularly
## /tmpdata - very fast [1mlocal(B[m space, no backups, no guarantees
## /scratch - fast [1mcluster(B[m space, no backups, no guarantees
## 
## Find a tool: module spider &lt;search_term&gt;
## Load a tool: module load &lt;module_name&gt;
## 
## Should you have any questions, please contact GBCS (gbcs@embl.de)
## or have a look here: https://gbcs.embl.de
## 
## mkdir: cannot create directory â€˜/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255/withzerosâ€™: No such file or directory
## mkdir: cannot create directory â€˜/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255/mapHighâ€™: No such file or directory
## zsh:15: no matches found: /Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255/*.bg
</code></pre>
<h1 id="part-iv-joining-tables">Part IV : Joining tables</h1>
<h3 id="protein-coding-genes_2">Protein coding genes</h3>
<pre><code class="language-r">main_directory &lt;- &quot;/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Cugusi2022/dTAG7/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/merged/stranded/bedgraph255&quot;
write_file_protein_coding &lt;- paste0(main_directory,&quot;/protCoding_dTAG_Cugusi_stranded_20230810.tsv&quot;)
write_file_lncRNA &lt;- paste0(main_directory,&quot;/lncRNA_dTAG_Cugusi_stranded_20230810.tsv&quot;)
Big_tsv &lt;-paste0(main_directory,&quot;/dTAG_Cugusi_stranded_20230810.tsv&quot;)  
</code></pre>
<pre><code class="language-r">main_directory &lt;- &quot;/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/Xiong2021/analysis_KD/bedgraph&quot;
write_file_protein_coding &lt;- paste0(main_directory,&quot;/protCoding_XiongsiSAFB_stranded_20240214.tsv&quot;)
write_file_lncRNA &lt;- paste0(main_directory,&quot;/lncRNA_XiongsiSAFB_stranded_20240214.tsv&quot;)
Big_tsv &lt;-paste0(main_directory,&quot;/XiongsiSAFB_stranded_20240214.tsv&quot;)  
</code></pre>
<pre><code class="language-r">main_directory &lt;- &quot;/Volumes/cristo-nas-shared/Personal_documents/Victor/DATA/SousaLuis2021/FASTQ_FILES/folder_clean/BAM/deduplicated/MAPQ255/stranded/bedgraph255&quot;
write_file_protein_coding &lt;- paste0(main_directory,&quot;/protCoding_SousaLuis_stranded_20240322.tsv&quot;)
write_file_lncRNA &lt;- paste0(main_directory,&quot;/lncRNA_SousaLuis_stranded_20240322.tsv&quot;)
Big_tsv &lt;-paste0(main_directory,&quot;/SousaLuis_stranded_20240322.tsv&quot;)  
</code></pre>
<p>~5 min</p>
<pre><code class="language-r">library(dplyr)
</code></pre>
<pre><code>## 
## Attaching package: 'dplyr'
</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag
</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union
</code></pre>
<pre><code class="language-r">library(purrr)

### DO NOT FORGET TO CHANGE THE BEDGRAPH EXTENSION TO TH ONE IN USE IN THE FOLDER
window &lt;- 200

working_directory &lt;- paste0(main_directory,&quot;/protein_coding_score&quot;)
bedgraph_files &lt;- list.files(main_directory, pattern = &quot;*.bg&quot;, full.names = TRUE)

files &lt;- bedgraph_files %&gt;%
  map(~{
    filename &lt;- tools::file_path_sans_ext(basename(.))
    file.path(working_directory, paste0(filename, &quot;.window&quot;, window, &quot;.MANE.wmean.name.score&quot;))
  })
# reading all the files
list_of_dfs &lt;- lapply(files, read.delim, header= F, sep= &quot;\t&quot;,  na.strings = &quot;NAN&quot;, dec = &quot;.&quot;, col.names = c(&quot;biotype&quot;,&quot;chr&quot;, &quot;coor1&quot;, &quot;coor2&quot;,&quot;transcript&quot;, &quot;gene&quot;, &quot;strand&quot;,&quot;window&quot;,&quot;id&quot;,&quot;dataset&quot;,&quot;score&quot;), stringsAsFactors=FALSE)
# joining all the files
joined_df &lt;- purrr::reduce(list_of_dfs, dplyr::left_join, by = c(&quot;biotype&quot;,&quot;chr&quot;,&quot;coor1&quot;,&quot;coor2&quot;,&quot;transcript&quot;,&quot;gene&quot;,&quot;strand&quot;,&quot;window&quot;,&quot;id&quot;)) %&gt;% 
   dplyr::filter(strand!=&quot;Y&quot;) ## the last filter remove the PAR genes (pseudoautosomal genes both in X and Y)
</code></pre>
<pre><code>## Error in `purrr::reduce()`:
## ! Must supply `.init` when `.x` is empty.
</code></pre>
<pre><code class="language-r"> write.table(joined_df, file = write_file_protein_coding, sep = &quot;\t&quot;, row.names = FALSE, col.names = FALSE, quote = F)
</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object 'joined_df' not found
</code></pre>
<pre><code class="language-r"> rm(list_of_dfs)
 rm(joined_df)
</code></pre>
<pre><code>## Warning in rm(joined_df): object 'joined_df' not found
</code></pre>
<h3 id="lncrna_2">lncRNA</h3>
<pre><code class="language-r">library(dplyr)
library(purrr)

### DO NOT FORGET TO CHANGE THE BEDGRAPH EXTENSION TO TH ONE IN USE IN THE FOLDER
window &lt;- 200
working_directory &lt;- paste0(main_directory,&quot;/lncRNA_score&quot;)

bedgraph_files &lt;- list.files(main_directory, pattern = &quot;*.bg&quot;, full.names = TRUE)

files &lt;- bedgraph_files %&gt;%
  map(~{
    filename &lt;- tools::file_path_sans_ext(basename(.))
    file.path(working_directory, paste0(filename, &quot;.window&quot;, window, &quot;.MANE.wmean.name.score&quot;))
  })
# reading all the files
list_of_dfs &lt;- lapply(files, read.delim, header= F, sep= &quot;\t&quot;,  na.strings = &quot;NAN&quot;, dec = &quot;.&quot;, col.names = c(&quot;biotype&quot;,&quot;chr&quot;, &quot;coor1&quot;, &quot;coor2&quot;,&quot;transcript&quot;, &quot;gene&quot;, &quot;strand&quot;,&quot;window&quot;,&quot;id&quot;,&quot;dataset&quot;,&quot;score&quot;), stringsAsFactors=FALSE)
# joining all the files
joined_df &lt;- purrr::reduce(list_of_dfs, dplyr::left_join, by = c(&quot;biotype&quot;,&quot;chr&quot;,&quot;coor1&quot;,&quot;coor2&quot;,&quot;transcript&quot;,&quot;gene&quot;,&quot;strand&quot;,&quot;window&quot;,&quot;id&quot;)) %&gt;% 
   dplyr::filter(strand!=&quot;Y&quot;) ## the last filter remove the PAR genes (pseudoautosomal genes both in X and Y)
</code></pre>
<pre><code>## Error in `purrr::reduce()`:
## ! Must supply `.init` when `.x` is empty.
</code></pre>
<pre><code class="language-r"> write.table(joined_df, file = write_file_lncRNA, sep = &quot;\t&quot;, row.names = FALSE, col.names = FALSE, quote = F)
</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object 'joined_df' not found
</code></pre>
<pre><code class="language-r">  rm(list_of_dfs)
 rm(joined_df)
</code></pre>
<pre><code>## Warning in rm(joined_df): object 'joined_df' not found
</code></pre>
<h2 id="merging-the-two-files">Merging the two files :</h2>
<pre><code class="language-r">library(dplyr)

tsv_files &lt;- list.files(main_directory, pattern = &quot;*.tsv&quot;, full.names = TRUE)

files &lt;- tsv_files %&gt;%
  map(~{
    filename &lt;- tools::file_path_sans_ext(basename(.))
    file.path(main_directory, paste0(filename, &quot;.tsv&quot;))
  })
# reading all the files
list_of_dfs &lt;- lapply(files, read.delim, header= F, sep= &quot;\t&quot;,  na.strings = &quot;NAN&quot;, dec = &quot;.&quot;, stringsAsFactors=FALSE)
# joining all the files
bound_df &lt;- purrr::reduce(list_of_dfs, dplyr::bind_rows)
</code></pre>
<pre><code>## Error in `purrr::reduce()`:
## ! Must supply `.init` when `.x` is empty.
</code></pre>
<pre><code class="language-r">write.table(bound_df, file = Big_tsv, sep = &quot;\t&quot;, row.names = FALSE, col.names = FALSE, quote = F)
</code></pre>
<pre><code>## Error in eval(expr, envir, enclos): object 'bound_df' not found
</code></pre>
<pre><code class="language-r">rm(list_of_dfs)
rm(bound_df)
</code></pre>
<pre><code>## Warning in rm(bound_df): object 'bound_df' not found
</code></pre>
</div>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/components/prism-core.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/prismjs@1.29.0/plugins/autoloader/prism-autoloader.min.js" defer></script>
</body>
</html>
