---
title: |
        "TepR:
        Transcription elongation profile in R"
author: |
        "Victor Billon, Nicolas Descostes, and Gael Cristofari"
output:
    BiocStyle::pdf_document: default
vignette: >
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteIndexEntry{tepr}
---


# Introduction

TepR (Transcription elongation profile in R) is a package for analyzing data
from nascent RNA sequencing data (TT-seq, mNET-seq, PRO-seq, etc.).
The general principle relies on calculating the cumulative signal of nascent RNA
sequencing over the gene body of any given gene or transcription unit. TePR can
identify transcription attenuation sites by comparing profile to a null model
which assumes uniform read density over the entirety of the transcriptional
unit. It can also identify increased or diminished transcription attenuation by
comparing two conditions. Besides rigorous statistical testing and high
sensitivity, a major feature of TepR is its ability to provide the elongation
pattern of each individual gene, including the position of the main attenuation
point when such a phenomenon occurs.
Using TepR, users can visualize and refine genome-wide aggregated analyses of
elongation patterns to robustly identify effects specific to subsets of genes.
These metrics are suitable for internal comparisons (between genes in each
condition) and for studying elongation of the same gene in different conditions
or comparing it to a perfect theoretical uniform elongation.

# Getting help

Please submit issues or questions on Github for a most efficient communication.

# Standard workflow

## Quick start

The example below shows a complete run of the package using two experiments,
the analysis is limited to 100 transcripts.

```{r quickstart, eval = FALSE}

library(tepr)

## Parameters
expthres <- 0.1
windsize <- 200

## Read input tables
expdf <- read.csv(system.file("extdata/exptab.csv"))
transdf <- read.delim(system.file(extdata/cugusi_100.tsv), header = FALSE)

## Calculate Average Expression and Filter Transcript Data
resallexprs <- averageandfilterexprs(expdf, transdf, expthres)

## Count NA values per transcript and condition
rescountna <- countna(resallexprs, expdf)

## Compute ECDF for Genes Based on Expression Data
resecdf <- genesECDF(resallexprs, expdf)

## Compute Mean and Differences of Scores for Each Condition
resmeandiff <- meandifference(resecdf, expdf, windsize)

## Split the results by transcripts
bytranslistmean <- split(resmeandiff, factor(resmeandiff$transcript))

## Calculate Area Under Curve (AUC) and Differences of AUC for Transcript Data
resauc <- allauc(bytranslistmean, expdf, nbwindows)
```
