---
title: "TepR: Transcription elongation profile in R"
author: "Victor Billon, Nicolas Descostes, and Gael Cristofari"
package: tepr
output:
    BiocStyle::pdf_document
vignette: >
    %\VignetteIndexEntry{tepr}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---


# Introduction

TepR (Transcription elongation profile in R) is a package for analyzing data
from nascent RNA sequencing data (TT-seq, mNET-seq, PRO-seq, etc.).
The general principle relies on calculating the cumulative signal of nascent RNA
sequencing over the gene body of any given gene or transcription unit. TePR can
identify transcription attenuation sites by comparing profile to a null model
which assumes uniform read density over the entirety of the transcriptional
unit. It can also identify increased or diminished transcription attenuation by
comparing two conditions. Besides rigorous statistical testing and high
sensitivity, a major feature of TepR is its ability to provide the elongation
pattern of each individual gene, including the position of the main attenuation
point when such a phenomenon occurs.
Using TepR, users can visualize and refine genome-wide aggregated analyses of
elongation patterns to robustly identify effects specific to subsets of genes.
These metrics are suitable for internal comparisons (between genes in each
condition) and for studying elongation of the same gene in different conditions
or comparing it to a perfect theoretical uniform elongation.

# Getting help

Please submit issues or questions on Github for a most efficient communication.

# Quick start

The example below shows a complete run of the package using two experiments,
the analysis is limited to 100 transcripts.

```{r quickstart, eval = FALSE}

library(tepr)

## Parameters
exppath <-  system.file("extdata", "exptab.csv", package="tepr")
transpath <- system.file("extdata", "cugusi_100.tsv", package="tepr")
expthres <- 0.1
windsize <- 200

## Read input tables
expdf <- read.csv(exppath)
transdf <- read.delim(transpath, header = FALSE)

## Calculate Average Expression and Filter Transcript Data
resallexprs <- averageandfilterexprs(expdf, transdf, expthres)

## Count NA values per transcript and condition
rescountna <- countna(resallexprs, expdf)

## Compute ECDF for Genes Based on Expression Data
resecdflist <- genesECDF(resallexprs, expdf)
resecdf <- resecdflist[[1]]
nbwindows <- resecdflist[[2]]

## Compute Mean and Differences of Scores for Each Condition
resmeandiff <- meandifference(resecdf, expdf, windsize)

## Split the results by transcripts
bytranslistmean <- split(resmeandiff, factor(resmeandiff$transcript))

## Calculate Area Under Curve (AUC) and Differences of AUC for Transcript Data
resauc <- allauc(bytranslistmean, expdf, nbwindows)

## Identify the Knee and Max ECDF Differences for Each Transcript
resknee <- kneeid(bytranslistmean, expdf)

## Calculate Attenuation from AUC and Other Transcript Features
resatt <- attenuation(resauc, resknee, rescountna, bytranslistmean, expdf,
    resmeandiff)

## Define Universe and Group of Genes Based on Expression Data
res <- universegroup(resatt)
```

# Data

## Description

For a complete analysis, we will use the data of Cugusi *et al* [REF]. To validate this strategy and explore the relevance of TepR metrics in uncovering gene-specific characteristics of elongation, we re-analyzed previous experiments showing that heat shock (HS) induces attenuation in human cultured cells (REF). We compared TT-seq data from HS-stressed and control MRC5-VA cells. The stressed cells were grown for 2h at 42°C, while control cells were kept at 37°C. Both samples underwent a 15-minute pulse of 4-thiouridine (4sU) labelling of nascent transcripts, followed by purification and sequencing.

## Download

The bedgraph files, mappability track and black list necessary for performing the entire analysis can be downloaded from zenodo:

```{bash raw-files, eval = FALSE}

TO DO
-----

#!/usr/bin/sh

mkdir data

wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.basic.annotation.gtf.gz -P data/ && gunzip data/gencode.v43.basic.annotation.gtf.gz
wget https://github.com/Boyle-Lab/Blacklist/blob/master/lists/hg38-blacklist.v2.bed.gz -P data/ && gunzip data/hg38-blacklist.v2.bed.gz

* k50.umap.hg38.0.8.bed (k50.umap.hg38.0.8.bed) - sent by victor
* ctrl_rep1.forward.bg (ctrl_rep1.forward.bg) - sent by victor
* ctrl_rep1.reverse.bg (ctrl_rep1.reverse.bg) - sent by victor
* ctrl_rep2.forward.bg (ctrl_rep2.forward.bg) - sent by victor
* ctrl_rep2.reverse.bg (ctrl_rep2.reverse.bg) - sent by victor
* HS_rep1.forward.bg (HS_rep1.forward.bg) - sent by victor
* HS_rep1.reverse.bg (HS_rep1.reverse.bg) - sent by victor
* HS_rep2.forward.bg (HS_rep2.forward.bg) - sent by victor
* HS_rep2.reverse.bg (HS_rep2.reverse.bg) - sent by victor

```

If one would like to skip the preprocessing, the result file can be downloaded with:

```{bash cugusi-table, eval = FALSE}

TO DO
```

## Pre-processing

To perform the pre-processing steps, we provide a conda environment that can be activated with:

```{bash conda, eval = FALSE}
TO DO
```

Retrieve the protein-coding and long non-coding RNA (lncRNA) annotations from the gencode file:

```{bash protcod-lncrna, eval = FALSE}
#!/usr/bin/sh

cd data/

## Building the protein-coding bed file
grep -w transcript gencode.v43.basic.annotation.gtf | grep -w MANE_Select > MANE_Select.protein_coding.gtf
awk 'OFS="\t" {print $0}' MANE_Select.protein_coding.gtf | tr -d '";' | sort -k1,1 -k2,2n  | awk -F \t -v OFS='\t' '{print $0}' | awk -v OFS="\t" '{ print $1,$4,$5,$12,$16,$7}' > MANE_Select.protein_coding.bed

## Building the lncRNA bed file
grep -w transcript gencode.v43.basic.annotation.gtf | grep -w lncRNA | grep -w Ensembl_canonical | grep -v not_best_in_genome_evidence | grep -v 'transcript_support_level "5"' | grep -v 'transcript_support_level "4"' > Ensembl_canonical_TSL123.lncRNA.gtf
awk 'OFS="\t" {print $0}' Ensembl_canonical_TSL123.lncRNA.gtf | tr -d '";' | sort -k1,1 -k2,2n  | awk -F \t -v OFS='\t' '{print $0}' | awk -v OFS="\t" '{ print $1,$4,$5,$12,$16,$7}' > Ensembl_canonical_TSL123.lncRNA.bed

rm MANE_Select.protein_coding.gtf Ensembl_canonical_TSL123.lncRNA.gtf
```

Remove the black list and make windows for protein-coding and lncRNA files:

```{bash remove, eval = FALSE}
#!/usr/bin/sh

WORKING="."
blacklist="hg38-blacklist.v2.bed"
window=200

mkdir makewindow

## Removing black list and computes windows for protein-coding annotations
ANNOTATION="MANE_Select.protein_coding.bed"
awk -F "\t" -v OFS="\t" '{print $1,$2,$3,$4"_"$5"_"$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v  | bedtools makewindows -n $window -i srcwinnum -b stdin | sort -k1,1 -k2,2n > $WORKING/makewindow/v43.MANE_protein.window${window}.bed
echo $WORKING/makewindow/v43.MANE_protein.window${window}.bed

## Removing black list and computes windows for lncRNA annotations
ANNOTATION="Ensembl_canonical_TSL123.lncRNA.bed"
awk -F "\t" -v OFS="\t" '{print $1,$2,$3,$4"_"$5"_"$6}' $ANNOTATION | bedtools intersect -a stdin -b $blacklist -v  | bedtools makewindows -n $window -i srcwinnum -b stdin | sort -k1,1 -k2,2n > $WORKING/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed
echo $WORKING/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed
```