####################
# This script aims at ensuring the conformity of the R code with the bash one.
#
# Descostes - July 2024 - R-4.4.1
####################



library("GenomeInfoDb")
library("GenomicRanges")

source("commons.R")


##################
# PARAMETERS
##################

## Files used for R and bash
maptrackpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/k50.Unique.Mappability.bed" # nolint
exptabpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/exptab.csv"
nbcpu <- 6
database_name <- "org.Hs.eg.db"

## Files obtained with BashAndR/pre-study/preprocessing.sh
protcodbedshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/MANE_Select.protein_coding.bed" # nolint
lncrnabedshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/Ensembl_canonical_TSL123.lncRNA.bed" # nolint
protcodbednoblackwindshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/makewindow/v43.MANE_protein.window200.bed" # nolint
lncrnanednoblackwindshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed" # nolint
blacklistshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/hg38-blacklist.v2.bed" # nolint
protcodnoblackfromshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/tmp2.bed" # nolint
lncrnanoblackfromshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/tmp4.bed" # nolint

## R objects generated by R/pre-study/preprocessing.R
robjoutputfold <- "/g/romebioinfo/Projects/tepr/robjsave"
robjlist <- list(protcodbed = file.path(robjoutputfold, "protcodbed.rds"),
    protcodgr = file.path(robjoutputfold, "protcodgr.rds"),
    lncrnabed = file.path(robjoutputfold, "lncrnabed.rds"),
    lncrnagr = file.path(robjoutputfold, "lncrnagr.rds"))

## R objects generated by BashAndR/pre-study/test-preprocessing.R
robjlisttestpre <- list(joined_df = file.path(robjoutputfold, "joined_df.rds"),
    joined_df_lncrna = file.path(robjoutputfold, "joined_df_lncrna.rds"),
    bound_df = file.path(robjoutputfold, "bound_df.rds")) ## final df of scores

## other params
windsize <- 200


##################
#FUNCTIONS
##################


verifybed <- function(bed1, bed2, nbcol = 6, includeposition = TRUE) {

    if (!isTRUE(all.equal(nrow(bed1), nrow(bed2))))
        stop("bed1 and bed2 have different nb of rows")

    message("Building strings for comparison in verifybed")
    if (includeposition) {
        bedstr1 <- paste(bed1[, 1], bed1[, 2], bed1[, 3], bed1[, 4], bed1[, 5], if(nbcol == 6) bed1[, 6], sep="-") # nolint
        bedstr2 <- paste(bed2[, 1], bed2[, 2], bed2[, 3], bed2[, 4], bed2[, 5], if(nbcol == 6) bed2[, 6], sep="-") # nolint
        idxcheckvec <- seq_len(5)
    } else {
        bedstr1 <- paste(bed1[, 1], bed1[, 4], bed1[, 5], if(nbcol == 6) bed1[, 6], sep="-") # nolint
        bedstr2 <- paste(bed2[, 1], bed2[, 4], bed2[, 5], if(nbcol == 6) bed2[, 6], sep="-") # nolint
        idxcheckvec <- c(1, 4, 5, 6)
    }

    idx <- match(bedstr1, bedstr2)
    idxna <- which(is.na(idx))
    lna <- length(idxna)
    if (!isTRUE(all.equal(lna, 0)))
        stop("The gene symbols-chrom are different")
    bed2 <- bed2[idx, ]
    invisible(sapply(idxcheckvec, function(i) {
        idx <- which(bed1[, i] != bed2[, i])
        lidx <- length(idx)
        if (!isTRUE(all.equal(lidx, 0)))
            stop("Difference in col ", i)
    }))
}

removepary <- function(dfbed) {
    idx <- grep("_PAR_Y", dfbed[, 4])
    if (!isTRUE(all.equal(length(idx), 0))) {
        message("\t\t Remove PARY")
        dfbed[idx, 4] <- gsub("_PAR_Y", "", dfbed[idx, 4])
    }
    return(dfbed)
}

comparenoblack <- function(bashpath, dfbed) {
    ## Read file obtained with bash
    fromsh <- read.delim(bashpath, header = FALSE)

    ## Remove suffix "_PAR_Y" if present in dfbed
    dfbed <- removepary(dfbed)

    ## Remove last column and add transcript names and strand to match the robj
    ## format
    tmplist <- strsplit(fromsh[, 4], "_")
    tmpnames <- sapply(tmplist, "[", 1)
    tmpstrand <- sapply(tmplist, function(x) x[length(x)])
    fromsh <- data.frame(fromsh$V1, fromsh$V2, fromsh$V3, tmpnames, tmpstrand)

    verifybed(dfbed, fromsh, nbcol = 5)
    return(list(fromsh, dfbed))
}

grtobed <- function(grobj) {
    res <- data.frame(GenomeInfoDb::seqnames(grobj), BiocGenerics::start(grobj),
        BiocGenerics::end(grobj), names(grobj),
        strand = BiocGenerics::strand(grobj))
    return(res)
}

separateframe <- function(dfbed) {

        reslist <- strsplit(dfbed[, 4], "_")
        trsnamevec <- sapply(reslist,  "[", 1)

        if (isTRUE(all.equal(ncol(dfbed), 5))) { # bed from r

            framevec <- sapply(reslist, "[", 2)
            strandvec <- dfbed[, 5]

            ## Makes framevec 1-based and numeric
            framevec <- as.numeric(gsub("frame", "", framevec))
            framevec[which(is.na(framevec))] <- 0
            framevec <- framevec + 1
        } else {
            framevec <- as.numeric(sapply(reslist, function(x) x[length(x)]))
            strandvec <- sapply(reslist, function(x) x[length(x) - 1])
        }

        ## Verify there is one frame per transcript line
        if (!isTRUE(all.equal(length(trsnamevec), length(framevec))))
            stop("Problem of correspondance between trsnamevec and framevec")

        ## Separate transcript names from frame in two columns 
        resbed <- data.frame(chrom = dfbed[, 1], start = dfbed[, 2],
            end = dfbed[, 3], name = trsnamevec, strand = strandvec,
            frame = framevec)
        return(resbed)
}

comparewind <- function(fromr_noblackshgr, fromsh_noblackwindpath, windsize) {
    ## Preparing bed df
    message("Creating ", windsize, " bins for r object")
    fromr_windgr <- makewindowsbedtools(fromr_noblackshgr, windsize)
    fromr_windbed <- grtobed(fromr_windgr)
    message("Reading ", windsize, " bins for sh object")
    fromsh_windbed <- read.delim(fromsh_noblackwindpath, header = FALSE)

    ## Remove suffix "_PAR_Y" if present in bed
    message("Removing suffixes from bed data.frame")
    fromr_windbed <- removepary(fromr_windbed)
    fromsh_windbed <- removepary(fromsh_windbed)

    ## Separate transcript names from frame in two columns
    message("Formatting bed columns")
    fromr_windbed <- separateframe(fromr_windbed)
    fromsh_windbed <- separateframe(fromsh_windbed)

    ## Note: the coordinates of the bins are not exactly the same between the
    ## R function "tile" and the bedtools function "makewindow". Therefore,
    ## verifybed(fromr_windbed, fromsh_windbed) returns an error.
    ## However, when the start and end column are left aside when comparing the
    ## two data.frame, it gives an equality. This means that each method gives
    ## the same number of bins but with slightly different coordinates.
    verifybed(fromr_windbed, fromsh_windbed, includeposition = FALSE)
    return(fromr_windgr)
}




##################
# MAIN
##################

message("If no error messages appear, this means conformity.")

## Loading objects
objlist <- mapply(function(obj, nameobj) {
    message("Loading ", nameobj)
    res <- readRDS(obj)
    return(res)
}, robjlist, names(robjlist), SIMPLIFY = FALSE, USE.NAMES = TRUE)

## Compare the bed files before removing black lists
protcodbedsh <- read.delim(protcodbedshpath, header = FALSE)
lncrnabedsh <- read.delim(lncrnabedshpath, header = FALSE)
verifybed(objlist[["protcodbed"]], protcodbedsh)
verifybed(objlist[["lncrnabed"]], lncrnabedsh)

## Exclude black list with the file that was used in bash
## NOTE: In the current code (preprocessing) the black list is retrieved from
## a database.
blacklistsh <- read.delim(blacklistshpath, header = FALSE)
blacklistshgr <- bedtogr(blacklistsh, strand = FALSE)
protcodnoblackshgr <- excludeorkeepgrlist(objlist[["protcodgr"]], blacklistshgr)
lncrnanoblackshgr <- excludeorkeepgrlist(objlist[["lncrnagr"]], blacklistshgr)
protcodnoblacksh <- grtobed(protcodnoblackshgr)
lncrnanoblacksh <- grtobed(lncrnanoblackshgr)

## Compare coord with black list removed to the files obtained with bash
res1 <- comparenoblack(protcodnoblackfromshpath, protcodnoblacksh)
res2 <- comparenoblack(lncrnanoblackfromshpath, lncrnanoblacksh)

## Compare windows coordinates with bash files
protcodwindows <- comparewind(protcodnoblackshgr, protcodbednoblackwindshpath,
    windsize)
lncrnawindows <- comparewind(lncrnanoblackshgr, lncrnanednoblackwindshpath,
    windsize)

## Build the data.frame with all the info starting from
## protcodnoblackshgr/lncrnanoblackshgr, removing the low mappability track,
## then calling buildscoreforintervals. It has to be compared to the bigTSV in
## the bash script
maptrack <- read.delim(maptrackpath, header = FALSE)
maptrackgr <- bedtogr(maptrack)
protcodnoblacknomapgr <- excludeorkeepgrlist(protcodnoblackshgr, maptrackgr,
    removefrom = FALSE)
lncrnanoblacknomapgr <- excludeorkeepgrlist(lncrnanoblackshgr, maptrackgr,
    removefrom = FALSE)

## Make windows of windsize for each annotation
protcodwindows <- makewindowsbedtools(protcodnoblacknomapgr, windsize)
lncrnawindows <- makewindowsbedtools(lncrnanoblacknomapgr, windsize)

## Retrieving values from bigwig files
exptab <- read.csv(exptabpath, header = TRUE)
protcoddf <- buildscoreforintervals(protcodwindows, exptab, "protein_coding",
    nbcpu, database_name)
