####################
# This script aims at ensuring the conformity of the R code with the bash one.
#
# Descostes - July 2024 - R-4.4.1
####################



library("GenomeInfoDb")
library("GenomicRanges")

source("commons.R")


##################
# PARAMETERS
##################

## Files obtained with BashAndR/pre-study/preprocessing.sh
protcodbedshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/MANE_Select.protein_coding.bed" # nolint
lncrnabedshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/Ensembl_canonical_TSL123.lncRNA.bed" # nolint
protcodbednoblackwindshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/makewindow/v43.MANE_protein.window200.bed" # nolint
lncrnanednoblackwindshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/makewindow/v43.Ensembl_canonical_TSL123.lncRNA.bed" # nolint
blacklistshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/hg38-blacklist.v2.bed" # nolint
protcodnoblackfromshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/tmp2.bed" # nolint
lncrnanoblackfromshpath <- "/g/romebioinfo/Projects/tepr/downloads/annotations/tmp4.bed" # nolint

## R objects generated by R/pre-study/preprocessing.R
robjoutputfold <- "/g/romebioinfo/Projects/tepr/robjsave"
robjlist <- list(protcodbed = file.path(robjoutputfold, "protcodbed.rds"),
    protcodgr = file.path(robjoutputfold, "protcodgr.rds"),
    lncrnabed = file.path(robjoutputfold, "lncrnabed.rds"),
    lncrnagr = file.path(robjoutputfold, "lncrnagr.rds"))

## R objects generated by BashAndR/pre-study/test-preprocessing.R
robjlisttestpre <- list(joined_df = file.path(robjoutputfold, "joined_df.rds"),
    joined_df_lncrna = file.path(robjoutputfold, "joined_df_lncrna.rds"),
    bound_df = file.path(robjoutputfold, "bound_df.rds")) ## final df of scores

## other params
windsize <- 200


##################
#FUNCTIONS
##################


verifybed <- function(bed1, bed2, nbcol = 6, includeposition = TRUE) {

    if (!isTRUE(all.equal(nrow(bed1), nrow(bed2))))
        stop("bed1 and bed2 have different nb of rows")

    message("Building strings for comparison in verifybed")
    if (includeposition) {
        bedstr1 <- paste(bed1[, 1], bed1[, 2], bed1[, 3], bed1[, 4], bed1[, 5], if(nbcol == 6) bed1[, 6], sep="-") # nolint
        bedstr2 <- paste(bed2[, 1], bed2[, 2], bed2[, 3], bed2[, 4], bed2[, 5], if(nbcol == 6) bed2[, 6], sep="-") # nolint
        idxcheckvec <- seq_len(5)
    } else {
        bedstr1 <- paste(bed1[, 1], bed1[, 4], bed1[, 5], if(nbcol == 6) bed1[, 6], sep="-") # nolint
        bedstr2 <- paste(bed2[, 1], bed2[, 4], bed2[, 5], if(nbcol == 6) bed2[, 6], sep="-") # nolint
        idxcheckvec <- c(1, 4, 5, 6)
    }

    idx <- match(bedstr1, bedstr2)
    idxna <- which(is.na(idx))
    lna <- length(idxna)
    if (!isTRUE(all.equal(lna, 0)))
        stop("The gene symbols-chrom are different")
    bed2 <- bed2[idx, ]
    invisible(sapply(idxcheckvec, function(i) {
        idx <- which(bed1[, i] != bed2[, i])
        lidx <- length(idx)
        if (!isTRUE(all.equal(lidx, 0)))
            stop("Difference in col ", i)
    }))
}

removepary <- function(dfbed) {
    idx <- grep("_PAR_Y", dfbed[, 4])
    if (!isTRUE(all.equal(length(idx), 0))) {
        message("\t\t Remove PARY")
        dfbed[idx, 4] <- gsub("_PAR_Y", "", dfbed[idx, 4])
    }
    return(dfbed)
}

comparenoblack <- function(bashpath, dfbed) {
    ## Read file obtained with bash
    fromsh <- read.delim(bashpath, header = FALSE)

    ## Remove suffix "_PAR_Y" if present in dfbed
    dfbed <- removepary(dfbed)

    ## Remove last column and add transcript names and strand to match the robj
    ## format
    tmplist <- strsplit(fromsh[, 4], "_")
    tmpnames <- sapply(tmplist, "[", 1)
    tmpstrand <- sapply(tmplist, function(x) x[length(x)])
    fromsh <- data.frame(fromsh$V1, fromsh$V2, fromsh$V3, tmpnames, tmpstrand)

    verifybed(dfbed, fromsh, nbcol = 5)
    return(list(fromsh, dfbed))
}

grtobed <- function(grobj) {
    res <- data.frame(GenomeInfoDb::seqnames(grobj), BiocGenerics::start(grobj),
        BiocGenerics::end(grobj), names(grobj),
        strand = BiocGenerics::strand(grobj))
    return(res)
}

separateframe <- function(dfbed) {

        reslist <- strsplit(dfbed[, 4], "_")
        trsnamevec <- sapply(reslist,  "[", 1)

        if (isTRUE(all.equal(ncol(dfbed), 5))) { # bed from r

            framevec <- sapply(reslist, "[", 2)
            strandvec <- dfbed[, 5]

            ## Makes framevec 1-based and numeric
            framevec <- as.numeric(gsub("frame", "", framevec))
            framevec[which(is.na(framevec))] <- 0
            framevec <- framevec + 1
        } else {
            framevec <- as.numeric(sapply(reslist, function(x) x[length(x)]))
            strandvec <- sapply(reslist, function(x) x[length(x) - 1])
        }

        ## Verify there is one frame per transcript line
        if (!isTRUE(all.equal(length(trsnamevec), length(framevec))))
            stop("Problem of correspondance between trsnamevec and framevec")

        ## Separate transcript names from frame in two columns 
        resbed <- data.frame(chrom = dfbed[, 1], start = dfbed[, 2],
            end = dfbed[, 3], name = trsnamevec, strand = strandvec,
            frame = framevec)
        return(resbed)
}

comparewind <- function(fromr_noblackshgr, fromsh_noblackwindpath, windsize) {
    ## Preparing bed df
    message("Creating ", windsize, " bins for r object")
    fromr_windgr <- makewindowsbedtools(fromr_noblackshgr, windsize)
    fromr_windbed <- grtobed(fromr_windgr)
    message("Reading ", windsize, " bins for sh object")
    fromsh_windbed <- read.delim(fromsh_noblackwindpath, header = FALSE)

    ## Remove suffix "_PAR_Y" if present in bed
    message("Removing suffixes from bed data.frame")
    fromr_windbed <- removepary(fromr_windbed)
    fromsh_windbed <- removepary(fromsh_windbed)

    ## Separate transcript names from frame in two columns
    message("Formatting bed columns")
    fromr_windbed <- separateframe(fromr_windbed)
    fromsh_windbed <- separateframe(fromsh_windbed)

    ## Note: the coordinates of the bins are not exactly the same between the
    ## R function "tile" and the bedtools function "makewindow". Therefore,
    ## verifybed(fromr_windbed, fromsh_windbed) returns an error.
    ## However, when the start and end column are left aside when comparing the
    ## two data.frame, it gives an equality. This means that each method gives
    ## the same number of bins but with slightly different coordinates.
    verifybed(fromr_windbed, fromsh_windbed, includeposition = FALSE)
}




##################
# MAIN
##################

message("If no error messages appear, this means conformity.")

## Loading objects
objlist <- mapply(function(obj, nameobj) {
    message("Loading ", nameobj)
    res <- readRDS(obj)
    return(res)
}, robjlist, names(robjlist), SIMPLIFY = FALSE, USE.NAMES = TRUE)

## Compare the bed files before removing black lists
protcodbedsh <- read.delim(protcodbedshpath, header = FALSE)
lncrnabedsh <- read.delim(lncrnabedshpath, header = FALSE)
verifybed(objlist[["protcodbed"]], protcodbedsh)
verifybed(objlist[["lncrnabed"]], lncrnabedsh)

## Exclude black list with the file that was used in bash
## NOTE: In the current code (preprocessing) the black list is retrieved from
## a database.
blacklistsh <- read.delim(blacklistshpath, header = FALSE)
blacklistshgr <- bedtogr(blacklistsh, strand = FALSE)
protcodnoblackshgr <- excludeorkeepgrlist(objlist[["protcodgr"]], blacklistshgr)
lncrnanoblackshgr <- excludeorkeepgrlist(objlist[["lncrnagr"]], blacklistshgr)
protcodnoblacksh <- grtobed(protcodnoblackshgr)
lncrnanoblacksh <- grtobed(lncrnanoblackshgr)

## Compare coord with black list removed to the files obtained with bash
res1 <- comparenoblack(protcodnoblackfromshpath, protcodnoblacksh)
res2 <- comparenoblack(lncrnanoblackfromshpath, lncrnanoblacksh)

## Compare windows coordinates with bash files
fromsh_protcodwindbed <- read.delim(protcodbednoblackwindshpath, header = FALSE)
fromsh_lncrnawindbed <- read.delim(lncrnanednoblackwindshpath, header = FALSE)
comparewind(protcodnoblackshgr, protcodbednoblackwindshpath, windsize)
comparewind(lncrnanoblackshgr, lncrnanednoblackwindshpath, windsize)




!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
check in above variables the columns compared to what is below and then merge
the two df. Save the object and compare it to the df_bound object

!! TODO: the last filter remove the PAR genes ie.e pseudoautosomal genes both
!! in X and Y by doing 'strand != "Y"'

!!The data frame should contain the columns
[1] "biotype"               "chr"                   "coor1"
 [4] "coor2"                 "transcript"            "gene"
 [7] "strand"                "window"                "id"
[10] "ctrl_rep1.plus"        "ctrl_rep1.plus_score"  "ctrl_rep1.minus"
[13] "ctrl_rep1.minus_score" "ctrl_rep2.plus"        "ctrl_rep2.plus_score"
[16] "ctrl_rep2.minus"       "ctrl_rep2.minus_score" "HS_rep1.plus"
[19] "HS_rep1.plus_score"    "HS_rep1.minus"         "HS_rep1.minus_score"
[22] "HS_rep2.plus"          "HS_rep2.plus_score"    "HS_rep2.minus"
[25] "HS_rep2.minus_score"

Here is an example:

[[8]]
         biotype  chr  coor1  coor2        transcript   gene strand window
1 protein-coding chr1 923923 924026 ENST00000616016.5 SAMD11      +      1
2 protein-coding chr1 924026 924129 ENST00000616016.5 SAMD11      +      2
3 protein-coding chr1 924129 924232 ENST00000616016.5 SAMD11      +      3
4 protein-coding chr1 924232 924335 ENST00000616016.5 SAMD11      +      4
5 protein-coding chr1 924335 924438 ENST00000616016.5 SAMD11      +      5
6 protein-coding chr1 924438 924541 ENST00000616016.5 SAMD11      +      6
                            id         dataset score
1 ENST00000616016.5_SAMD11_+_1 HS_rep2.reverse  <NA>
2 ENST00000616016.5_SAMD11_+_2 HS_rep2.reverse  <NA>
3 ENST00000616016.5_SAMD11_+_3 HS_rep2.reverse  <NA>
4 ENST00000616016.5_SAMD11_+_4 HS_rep2.reverse  <NA>
5 ENST00000616016.5_SAMD11_+_5 HS_rep2.reverse  <NA>
6 ENST00000616016.5_SAMD11_+_6 HS_rep2.reverse  <NA>

> lapply(list_of_dfs, function(x) unique(x$biotype))
[[1]]
[1] "protein-coding"

[[2]]
[1] "protein-coding"

[[3]]
[1] "protein-coding"

[[4]]
[1] "protein-coding"

[[5]]
[1] "protein-coding"

[[6]]
[1] "protein-coding"

[[7]]
[1] "protein-coding"

[[8]]
[1] "protein-coding"

After retrieving scores for all files we obtain the variable joined_df (see
joined_df path in parameters of R\pre-study\conformity_with_bash.R) with:

 biotype  chr  coor1  coor2        transcript   gene strand window
1 protein-coding chr1 923923 924026 ENST00000616016.5 SAMD11      +      1
2 protein-coding chr1 924026 924129 ENST00000616016.5 SAMD11      +      2
3 protein-coding chr1 924129 924232 ENST00000616016.5 SAMD11      +      3
4 protein-coding chr1 924232 924335 ENST00000616016.5 SAMD11      +      4
5 protein-coding chr1 924335 924438 ENST00000616016.5 SAMD11      +      5
6 protein-coding chr1 924438 924541 ENST00000616016.5 SAMD11      +      6
                            id         dataset.x  score.x         dataset.y
1 ENST00000616016.5_SAMD11_+_1 ctrl_rep1.forward 0.000000 ctrl_rep1.reverse
2 ENST00000616016.5_SAMD11_+_2 ctrl_rep1.forward 0.000000 ctrl_rep1.reverse
3 ENST00000616016.5_SAMD11_+_3 ctrl_rep1.forward 0.000000 ctrl_rep1.reverse
4 ENST00000616016.5_SAMD11_+_4 ctrl_rep1.forward 0.000000 ctrl_rep1.reverse
5 ENST00000616016.5_SAMD11_+_5 ctrl_rep1.forward 0.000000 ctrl_rep1.reverse
6 ENST00000616016.5_SAMD11_+_6 ctrl_rep1.forward 0.000000 ctrl_rep1.reverse
  score.y       dataset.x.x score.x.x       dataset.y.y score.y.y
1    <NA> ctrl_rep2.forward  0.000000 ctrl_rep2.reverse      <NA>
2    <NA> ctrl_rep2.forward  0.000000 ctrl_rep2.reverse      <NA>
3    <NA> ctrl_rep2.forward  0.000000 ctrl_rep2.reverse      <NA>
4    <NA> ctrl_rep2.forward  0.000000 ctrl_rep2.reverse      <NA>
5    <NA> ctrl_rep2.forward  0.000000 ctrl_rep2.reverse      <NA>
6    <NA> ctrl_rep2.forward  0.000000 ctrl_rep2.reverse      <NA>
    dataset.x.x.x score.x.x.x   dataset.y.y.y score.y.y.y dataset.x.x.x.x
1 HS_rep1.forward    0.000000 HS_rep1.reverse        <NA> HS_rep2.forward
2 HS_rep1.forward    0.000000 HS_rep1.reverse        <NA> HS_rep2.forward
3 HS_rep1.forward    0.000000 HS_rep1.reverse        <NA> HS_rep2.forward
4 HS_rep1.forward    0.000000 HS_rep1.reverse        <NA> HS_rep2.forward
5 HS_rep1.forward    0.000000 HS_rep1.reverse        <NA> HS_rep2.forward
6 HS_rep1.forward    0.000000 HS_rep1.reverse        <NA> HS_rep2.forward
  score.x.x.x.x dataset.y.y.y.y score.y.y.y.y
1      0.000000 HS_rep2.reverse          <NA>
2      0.000000 HS_rep2.reverse          <NA>
3      0.000000 HS_rep2.reverse          <NA>
4      0.000000 HS_rep2.reverse          <NA>
5      0.000000 HS_rep2.reverse          <NA>
6      0.000000 HS_rep2.reverse          <NA>

